safehttp/form.go

```diff
@@ -0,0 +1,253 @@
+// associated with paramKey, it will return the default value and set the Form
+// error field.
+func (f *Form) Bool(paramKey string, defaultValue bool) bool {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramKey]
+	if !ok {
+		f.err = fmt.Errorf("no value found for key %q", paramKey)
+		return defaultValue
+	}
+	switch vals[0] {
+	case "true":
+		return true
+	case "false":
+		return false
+	default:
+		f.err = fmt.Errorf("values of form parameter %q not a boolean", paramKey)
+	}
+	return false
+}
+
+func clearSlice(slicePtr interface{}) error {
+	switch vs := slicePtr.(type) {
+	case *[]string:
+		*vs = nil
+	case *[]int:
+		*vs = nil
+	case *[]float64:
+		*vs = nil
+	case *[]uint64:
+		*vs = nil
+	case *[]bool:
+		*vs = nil
+	default:
+		return fmt.Errorf("type not supported in Slice call: %T", vs)
+	}
+	return nil
+}
+
+// Slice checks whether key paramKey maps to any query or form parameters. If it
+// does, it will try to convert them to the type of slice elements slicePtr
+// points to. If there are no values associated with paramKey or type conversion
+// fails at any point, the Form error field will be set and slicePtr will point
+// to nil.
+func (f *Form) Slice(slicePtr interface{}, paramKey string) {
+	if f.err != nil {
+		f.err = clearSlice(slicePtr)
+	}
+	mapVals, ok := f.values[paramKey]
+	if !ok {
+		clearSlice(slicePtr)
+		f.err = fmt.Errorf("no value found for key %q", paramKey)
+	}
+	switch values := slicePtr.(type) {
+	case *[]string:
+		res := make([]string, 0, len(mapVals))
+		for _, x := range mapVals {
+			res = append(res, x)
+		}
+		*values = res
+	case *[]int:
+		res := make([]int, 0, len(mapVals))
+		for _, x := range mapVals {
+			x, err := strconv.Atoi(x)
+			if err != nil {
+				f.err = err
+				*values = nil
+				return
+			}
+			res = append(res, x)
+			*values = res
+		}
+	case *[]uint64:
+		res := make([]uint64, 0, len(mapVals))
+		for _, x := range mapVals {
+			x, err := strconv.ParseUint(x, 10, 0)
+			if err != nil {
+				f.err = err
+				*values = nil
+				slicePtr = values
+				return
+			}
+			res = append(res, x)
+		}
+		*values = res
+	case *[]float64:
+		res := make([]float64, 0, len(mapVals))
+		for _, x := range mapVals {
+			x, err := strconv.ParseFloat(x, 64)
+			if err != nil {
+				f.err = err
+				*values = nil
+				return
+			}
+			res = append(res, x)
+		}
+		*values = res
+	case *[]bool:
+		res := make([]bool, 0, len(mapVals))
+		for _, x := range mapVals {
+			switch x {
+			case "true":
+				res = append(res, true)
+			case "false":
+				res = append(res, false)
+			default:
+				f.err = fmt.Errorf("values of form parameter %q not a boolean", paramKey)
+				*values = nil
+				return
+			}
+		}
+		*values = res
+	default:
+		f.err = clearSlice(slicePtr)
+	}
+	return
+}
+
+// Err returns the value of the Form error field. This will be nil unless an
+// error occured while accessing a parsed form value.
+func (f *Form) Err() error {
+	return f.err
+}
```

Comment:
Wasn't the plan to return and clear? So you can easily do:

```
x := f.Int("x", 0)
if f.Err() != nil {
  // Well, x is not *that* important. Let's continue.
}
y := f.Int("y", 0)
if f.Err() != nil {
  panic("y is very important!")
}
```

I actually like to hear back from other folks about this.

Reply:
http://doc/1OHXUwC-sOK4Uk6-By7_lPW8ioJRydd7AuO4gIMnUjNU#bookmark=id.6ndng7y6lrs8

Reply:
Seeing that usecase, it would make sense.. will keep refactoring on hold until we make a decision on this